tmol/score/cartbonded/cartbonded_energy_term.py

```diff
@@ -0,0 +1,285 @@
+            return
+
+        # Get the subgraphs for this block type
+        lengths, angles, torsions, improper = self.find_subgraphs(
+            block_type.bond_indices, block_type
+        )
+        cart_subgraphs = numpy.asarray(lengths + angles + torsions + improper)
+        setattr(block_type, "cartbonded_subgraphs", cart_subgraphs)
+
+        # Fetch the params from the database, updating the atom id store if necessary
+        cartbonded_params = self.get_params_for_res(block_type.name)
+        setattr(block_type, "cartbonded_params", cartbonded_params)
+
+    def setup_packed_block_types(self, packed_block_types: PackedBlockTypes):
+        super(CartBondedEnergyTerm, self).setup_packed_block_types(packed_block_types)
+        if hasattr(packed_block_types, "cartbonded_subgraphs"):
+            assert hasattr(packed_block_types, "cartbonded_subgraph_offsets")
+            assert hasattr(packed_block_types, "cartbonded_max_subgraphs_per_block")
+            assert hasattr(packed_block_types, "cartbonded_params_hash_keys")
+            assert hasattr(packed_block_types, "cartbonded_params_hash_values")
+            return
+
+        # Aggregate the subgraphs and collect metadata
+        total_subgraphs = sum(
+            bt.cartbonded_subgraphs.shape[0]
+            for bt in packed_block_types.active_block_types
+        )
+        subgraphs = numpy.full((total_subgraphs, 4), -1, dtype=numpy.int32)
+        subgraph_offsets = []
+        offset = 0
+        max_subgraphs_per_block = 0
+        for block_type in packed_block_types.active_block_types:
+            subgraph_offsets.append(offset)
+
+            n_subgraphs = block_type.cartbonded_subgraphs.shape[0]
+            subgraphs[offset : offset + n_subgraphs] = block_type.cartbonded_subgraphs
+            offset += n_subgraphs
+
+            max_subgraphs_per_block = max(
+                max_subgraphs_per_block, offset - subgraph_offsets[-1]
+            )
+        subgraphs = torch.from_numpy(subgraphs).to(device=self.device)
+        subgraph_offsets = numpy.asarray(subgraph_offsets, dtype=numpy.int32)
+        subgraph_offsets = torch.from_numpy(subgraph_offsets).to(device=self.device)
+        setattr(packed_block_types, "cartbonded_subgraphs", subgraphs)
+        setattr(packed_block_types, "cartbonded_subgraph_offsets", subgraph_offsets)
+        setattr(
+            packed_block_types,
+            "cartbonded_max_subgraphs_per_block",
+            max_subgraphs_per_block,
+        )
+
+        # Aggregate the params
+
+        # Calculate the total number of params
+        total_params = sum(
+            [len(bt.cartbonded_params) for bt in packed_block_types.active_block_types]
+        )
+        # get the params not associated with any specific residue
+        wildcard_params = self.get_params_for_res("wildcard").items()
+        total_params += len(wildcard_params)
+
+        # stupid hash function
+        def hash_fun(key, max_size):
```

Comment:
one potential source of error is that this needs to be synched w/ c++ version.  Maybe add a comment here about that (already 1 in cpp code).  Is it possible to have the magic numbers (0x1234, 3141) come from a single source?

Reply:
The hash function code itself also must be synced, so I was thinking it was better to abstract it at the functional level. I'm not sure what the best way to share the magic numbers would be. You could always do a text file, or pass these values in. Those seem kind of hacky, though.

Reply:
I've added a comment to the hash function here to address this.

Reply:
That's a great point, I did not think about putting python code in that directory. I see now that there are some examples of that.